### ISSUE TITLE
Pathological rebuild performance

### ISSUE THREAD

**Issue by meithecatte** [egraphs-good/egg#239]
Consider an analysis that is overwhelmingly likely to propagate updates throughout the e-graph. In my use-case, I was computing `AstSize` in my `Analysis`, as I needed it for some heuristics. The below example models this by always returning `DidMerge(true, true)`.

```rust
use egg::*;

// Runtime is extremely sensitive to this argument
const SIZE: usize = 14;

type EGraph = egg::EGraph<SymbolLang, MyAnalysis>;

struct MyAnalysis;

impl Analysis<SymbolLang> for MyAnalysis {
    type Data = ();

    fn make(_: &EGraph, _: &SymbolLang) {}
    fn merge(&mut self, _: &mut (), _: ()) -> DidMerge {
        DidMerge(true, true)
    }
}

fn wrap(egraph: &mut EGraph, mut v: Id) {
    for _ in 0..SIZE {
        let f = egraph.add(SymbolLang { op: "f".into(), children: vec![v] });
        let g = egraph.add(SymbolLang { op: "g".into(), children: vec![v] });
        egraph.union(f, g);
        v = f;
    }
}

fn main() {
    env_logger::init();

    let mut egraph = EGraph::new(MyAnalysis);
    let x = egraph.add(SymbolLang { op: "x".into(), children: vec![] });
    let y = egraph.add(SymbolLang { op: "y".into(), children: vec![] });
    wrap(&mut egraph, x);
    wrap(&mut egraph, y);
    egraph.rebuild();

    egraph.union(x, y);
    egraph.rebuild();
}
```

Running this with `RUST_LOG=egg=info cargo r --release` prints:

```
[2023-02-21T05:20:57Z INFO  egg::egraph] REBUILT! in 0.000s
      Old: hc size 58, eclasses: 30
      New: hc size 58, eclasses: 30
      unions: 0, trimmed nodes: 0
[2023-02-21T05:21:15Z INFO  egg::egraph] REBUILT! in 18.454s
      Old: hc size 58, eclasses: 29
      New: hc size 58, eclasses: 15
      unions: 14, trimmed nodes: 28
```

The issue is that `process_unions` iterates through `analysis_pending` in a random order, instead of bottom-up. Performing the latter accurately might be non-trivial, but while I haven't thought through this fully, I believe a very good approximation, that would recover the optimal asymptotic complexity, is to count how many times the analysis data of each e-node has been handled during the current rebuild, and in each iteration pick the one where this count is lowest (with something like `BinaryHeap`).

**Comment by meithecatte** [comment]
Hmm, I think that the counting solution won't be optimal - it'll probably avoid the exponential behavior, but in some cases it'll still be quadratic in the worst case.

**Comment by mwillsey** [comment]
I'm happy to take a fix here if you have one. Is this something you ran into in practice?

**Comment by meithecatte** [comment]
Yes, I ran into this after I tried including the minimum AstSize in my analysis. Most stuff worked fine, and then it encountered an e-graph that seemed to have hung during rebuilding. At first I thought I screwed up how another part of the analysis handles cycles.

**Comment by mwillsey** [comment]
Ah, interesting. Good catch.

I won't have time to get to this for a while, but I'm happy to merge a fix if you happen to make one.

**Comment by meithecatte** [comment]
I'm not sure what's the best approach here. One option I'm considering is to compute a topological sort of the e-graph at some point during the rebuild, and use that to guide the process of rebuilding the analysis data (picking the e-class that comes first in the topological order).

This ordering might become stale after the rebuild continues, but it should still be useful as a heuristic.

The question, then, is when to compute the toposort. One option would be the first time we reach the analysis update loop, but if we end up updating 3 nodes, that's obviously inefficient. So perhaps maintain a `HashMap` that remembers how many times we've updated the analysis for a node, and once it passes some small threshold, compute the toposort?

The issue then is to pick this threshold. The optimal choice will depend on how expensive the analysis is to compute, and thus will vary across applications. This makes me want to expose it as a tweakable knob, but on the other hand that is an entirely different game than avoiding rebuild times measured in days :upside_down_face: 

I suppose we'll have to see how introducing this toposort affects rebuild times in the typical case.

**Comment by Bastacyclop** [comment]
In my custom implementation of analyses, I used a queue instead of a stack for `analysis_pending`. Would that help here? 
- I have a Rust version used in my `egg-sketches` library https://github.com/Bastacyclop/egg-sketches/blob/main/src/analysis.rs ;
- and a Scala version used in the Rise language research project https://github.com/rise-lang/shine/blob/main/src/main/scala/rise/eqsat/Analysis.scala

**Comment by mwillsey** [comment]
I think switching to a queue seems reasonable, I'd merge such a change. @meithecatte do you think it's possible to check if that works on your workload?

**Comment by meithecatte** [comment]
I will have to find the relevant revision of my repo again, but check the following first: whereas my testcase iteratively expands `?a => (f ?a), (g ?a)` and unions the two together, try `?a => (f ?a), (g (h ?a))`.

**Comment by mwillsey** [comment]
I don't quite understand those rules. What's the `,` doing? Are those multipatterns?

**Comment by meithecatte** [comment]
I'd like to note, that if O(n²) performance, as above, is acceptable, then the following rebuilding algorithm can be proven to have this complexity in the worst case, in the case where there are no cycles:

> Throughout the timespan of one rebuild, track the number of times we've recomputed the analysis for a particular node. At each step, pick from the set of nodes we know are "dirty" the node with the lowest value of this count. Ties can be broken as desired.

**Proof sketch**: since we're assuming that there are no cycles, consider the topological sort `topo[1..=n]`, numbered in the direction where `topo[1]` is not the parent of any node. Note that once a prefix of the toposort doesn't have any elements in the dirty set, it'll stay that way. Inductively prove that the analysis of `topo[i]` will be computed at most `i` times, and that when this first happens for a particular `i`, all the other counters are also at most `i`.

If there *are* cycles, I don't think we can guarantee much, apart from termination with reasonable assumptions about the algebraic properties of the analysis itself. However from what I can see when thinking through an example with a cycle, it seems to me that both "least-recomputed" and "simple queue" handle this with similar performance characteristics.

Now that I think of it, though, it might be the case that the "simple queue" algorithm can also be proven to be worst-case quadratic with these assumptions. I'll have to think about this further.

**Comment by mwillsey** [comment]
Nice! So the solution is really just use a queue? @Bastacyclop would you like to file a PR?

**Comment by mwillsey** [comment]
I tried a simpler solution in #240, could y'all check if it meets your needs?

**Comment by yihozhang** [comment]
I think there is a worst-case O(n) algorithm for e-class analysis, which is to first run the rebuilding to fixpoint without doing any e-class analysis, at the same time tracking what set of old e-classes each new e-class corresponds to. We can then use this information to do an amortized maintenance of the e-class analysis in one scan. This works since rebuilding does not require an up-to-date e-class analysis

**Comment by meithecatte** [comment]
Do note that running the analysis is allowed to introduce new unions.

**Comment by yihozhang** [comment]
Oh nevermind, this might not work
